soroban-env-host/src/host/prng.rs

```diff
@@ -0,0 +1,151 @@
+///     can a caller control the seed for a "random" callee (since they can't
+///     observe the state of the "base" PRNG).
+///
+///   - Users _can_ reseed their frame-local PRNG if they want, which is a
+///     useful building block for random-commitment schemes. In particular if a
+///     contract is trying to make a "single random decision", and avoid having
+///     callers retry that decision repeatedly while aborting the transaction on
+///     any random decision the caller doesn't like, the contract can operate as
+///     a state machine like so:
+///
+///       - tx1: write commitment finalizing all inputs to "random action", plus
+///              N = current ledger and S = prng_new_bytes(32).
+///
+///       - tx2: re-read all committed values, if ledger > N, prng_reseed(S),
+///              and use PRNG to take "random" action committed-to.
+///
+///     With this design, assuming the contract does not expose any _online_
+///     method for its caller to observe the commitment it made in tx1, the
+///     caller (situated in the same transaction) won't know from its position
+///     whether it's to its advantage or not to abort tx1, so will naturally let
+///     it commit. Once the commitment is saved, it includes a "locked in"
+///     choice of seed, essentially propagating PRNG state out of a context the
+///     caller might have been able to influence its state via selective aborts
+///     (but didn't know whether to) into a context where the caller can no
+///     longer influence its state. The contract can then be re-invoked in the
+///     next ledger to load and execute the commitment, in tx2.
+///
+///   - We also include 3 building blocks for _using_ the PRNG: one basic one
+///     that just "generates a random BytesObject" (that the user can do
+///     anything they like with, including copying to guest linear memory), and
+///     two slightly more subtle but very standard operations that are easy to
+///     get wrong: an inclusive-range uniform u64 sampler and a Fisher-Yates
+///     vector shuffle. The latter is also hard to do cheaply in guest code
+///     without copying the vector into the guest and copying it back.
+///
+///   - All these PRNGs are ChaCha20: a strong, cheap, standard CSPRNG.
+///
+#[derive(Debug, Clone)]
+pub(crate) struct Prng(ChaCha20Rng);
+
+pub type Seed = <rand_chacha::ChaCha20Rng as rand::SeedableRng>::Seed;
+pub const SEED_BYTES: usize = core::mem::size_of::<Seed>();
+static_assertions::const_assert_eq!(SEED_BYTES, 32);
+
+impl Default for Prng {
+    fn default() -> Self {
+        Self(ChaCha20Rng::from_seed([0; SEED_BYTES]))
+    }
+}
+
+impl Prng {
+    fn charge_prng_bytes(&self, budget: &Budget, count: u64) -> Result<(), HostError> {
+        // TODO: add a ContractCostType for drawing PRNG bytes
+        Ok(())
+    }
+
+    pub(crate) fn reseed(&mut self, seed: Seed) {
+        self.0 = ChaCha20Rng::from_seed(seed);
```

Comment:
Do we need to do anything wrt requirements wrt `from_seed`  (I think not because we're using a cryptographic hash function, but checking)

```
PRNG implementations are allowed to assume that bits in the seed are well distributed. That means usually that the number of one and zero bits are roughly equal, and values like 0, 1 and (size - 1) are unlikely.
```

Reply:
From what I can find on the internet, proper function of ChaCha20 is not sensitive to bit-bias in its seed -- the seed just has to be (obviously) secret and not predictable. ChaCha20 is a minor variant of Salsa20, which was built to demonstrate that one can construct a viable stream cipher (and everything else cryptographic) out of a cryptographic hash, and as such its design is just a CHF of simple set of state-words: some fixed constants, the "key" ("seed" in our case) words, a stream identifier number, and a nonce that increments on each byte emitted. Changing one bit in any of these should do the same thing it does in any CHF: unpredictably flip half the bits. The stream and nonce words are actually _assumed_ to be highly biased numbers (mostly zeroes) and the mixing function is designed to integrate bits uniformly from all words in its input, as far as I can tell. See [section 4 of the design docs](https://cr.yp.to/snuffle/design.pdf) for some elaboration.

That said, I am not a cryptographer and I would be happy to add a documentation qualifier here that advises users to only ever use seeds that they got from the CSPRNG in the first place, since other seeds are likely _predictable_, which defeats the whole point. But I don't think we need to do anything extra here.

And yes, the seed _we_ supply (the one driving the outer `base_prng`) come from a CHF, so they should also satisfy this criterion already.